File: backend/controllers/parent.controller.js
import { getFirestore } from '../config/firebase.js';

export async function getAllParents(req, res) {
  try {
    const db = getFirestore();
    const snapshot = await db.collection('parents').get();
    const parents = snapshot.docs.map((doc) => ({ id: doc.id, ...doc.data() }));
    res.json({ success: true, data: parents });
  } catch (error) {
    res.status(500).json({ success: false, message: 'Failed to fetch parents' });
  }
}

export async function getParentById(req, res) {
  try {
    const db = getFirestore();
    const doc = await db.collection('parents').doc(req.params.id).get();
    if (!doc.exists) {
      return res.status(404).json({ success: false, message: 'Parent not found' });
    }
    res.json({ success: true, data: { id: doc.id, ...doc.data() } });
  } catch (error) {
    res.status(500).json({ success: false, message: 'Failed to fetch parent' });
  }
}

export async function createParent(req, res) {
  try {
    const db = getFirestore();
    const parentData = {
      ...req.body,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
    const docRef = await db.collection('parents').add(parentData);
    res.status(201).json({ success: true, data: { id: docRef.id, ...parentData } });
  } catch (error) {
    res.status(500).json({ success: false, message: 'Failed to create parent' });
  }
}

export async function updateParent(req, res) {
  try {
    const db = getFirestore();
    const updateData = { ...req.body, updatedAt: new Date().toISOString() };
    await db.collection('parents').doc(req.params.id).update(updateData);
    res.json({ success: true, message: 'Parent updated successfully' });
  } catch (error) {
    res.status(500).json({ success: false, message: 'Failed to update parent' });
  }
}

export async function deleteParent(req, res) {
  try {
    const db = getFirestore();
    await db.collection('parents').doc(req.params.id).delete();
    res.json({ success: true, message: 'Parent deleted successfully' });
  } catch (error) {
    res.status(500).json({ success: false, message: 'Failed to delete parent' });
  }
}

File: backend/controllers/parent_controller.py
from datetime import datetime, timezone
from flask import request, jsonify
from config.firebase_config import get_firestore


def _now():
    return datetime.now(timezone.utc).isoformat()


def get_all_parents():
    try:
        db = get_firestore()
        snapshot = db.collection('parents').get()
        parents = [{'id': doc.id, **doc.to_dict()} for doc in snapshot]
        return jsonify({'success': True, 'data': parents})
    except Exception:
        return jsonify({'success': False, 'message': 'Failed to fetch parents'}), 500


def get_parent_by_id(parent_id):
    try:
        db = get_firestore()
        doc = db.collection('parents').document(parent_id).get()
        if not doc.exists:
            return jsonify({'success': False, 'message': 'Parent not found'}), 404
        return jsonify({'success': True, 'data': {'id': doc.id, **doc.to_dict()}})
    except Exception:
        return jsonify({'success': False, 'message': 'Failed to fetch parent'}), 500


def create_parent():
    try:
        db = get_firestore()
        now = _now()
        parent_data = {**(request.get_json() or {}), 'createdAt': now, 'updatedAt': now}
        doc_ref = db.collection('parents').add(parent_data)[1]
        return jsonify({'success': True, 'data': {'id': doc_ref.id, **parent_data}}), 201
    except Exception:
        return jsonify({'success': False, 'message': 'Failed to create parent'}), 500


def update_parent(parent_id):
    try:
        db = get_firestore()
        update_data = {**(request.get_json() or {}), 'updatedAt': _now()}
        db.collection('parents').document(parent_id).update(update_data)
        return jsonify({'success': True, 'message': 'Parent updated successfully'})
    except Exception:
        return jsonify({'success': False, 'message': 'Failed to update parent'}), 500


def delete_parent(parent_id):
    try:
        db = get_firestore()
        db.collection('parents').document(parent_id).delete()
        return jsonify({'success': True, 'message': 'Parent deleted successfully'})
    except Exception:
        return jsonify({'success': False, 'message': 'Failed to delete parent'}), 500
